tests/check/check.go

```diff
@@ -196,3 +199,430 @@ func runTest(t *testing.T, test individualTest, params testParams, contextTupleT
+
+	ctx := context.Background()
+	resp, err := client.CreateStore(ctx, &openfgav1.CreateStoreRequest{Name: name})
+	require.NoError(t, err)
+	storeID := resp.GetId()
+
+	var modelID string
+	t.Run(name, func(t *testing.T) {
+		for _, stage := range matrix.Stages {
+			t.Run(fmt.Sprintf("stage_%s", stage.Name), func(t *testing.T) {
+				if stage.Model != "" {
+					model := testutils.MustTransformDSLToProtoWithID(stage.Model)
+					writeModelResponse, err := client.WriteAuthorizationModel(ctx, &openfgav1.WriteAuthorizationModelRequest{
+						StoreId:         storeID,
+						SchemaVersion:   schemaVersion,
+						TypeDefinitions: model.GetTypeDefinitions(),
+						Conditions:      model.GetConditions(),
+					})
+					require.NoError(t, err)
+					modelID = writeModelResponse.GetAuthorizationModelId()
+				}
+
+				tuples := stage.Tuples
+				tuplesLength := len(tuples)
+				if tuplesLength > 0 {
+					for i := 0; i < tuplesLength; i += writeMaxChunkSize {
+						end := int(math.Min(float64(i+writeMaxChunkSize), float64(tuplesLength)))
+						writeChunk := (tuples)[i:end]
+						_, err = client.Write(ctx, &openfgav1.WriteRequest{
+							StoreId:              storeID,
+							AuthorizationModelId: modelID,
+							Writes: &openfgav1.WriteRequestWrites{
+								TupleKeys: writeChunk,
+							},
+						})
+						require.NoError(t, err)
+					}
+				}
+
+				if len(stage.CheckAssertions) == 0 {
+					t.Skipf("no check assertions defined")
+				}
+				for _, assertion := range stage.CheckAssertions {
+					t.Run(fmt.Sprintf("assertion_%s", assertion.Name), func(t *testing.T) {
+						detailedInfo := fmt.Sprintf("Check request: %s. Tuples: %s. Contextual tuples: %s", assertion.Tuple, stage.Tuples, assertion.ContextualTuples)
+
+						var tupleKey *openfgav1.CheckRequestTupleKey
+						if assertion.Tuple != nil {
+							tupleKey = &openfgav1.CheckRequestTupleKey{
+								User:     assertion.Tuple.GetUser(),
+								Relation: assertion.Tuple.GetRelation(),
+								Object:   assertion.Tuple.GetObject(),
+							}
+						}
+						resp, err := client.Check(ctx, &openfgav1.CheckRequest{
+							StoreId:              storeID,
+							AuthorizationModelId: modelID,
+							TupleKey:             tupleKey,
+							ContextualTuples: &openfgav1.ContextualTupleKeys{
+								// TODO
+								TupleKeys: []*openfgav1.TupleKey{},
+							},
+							Context: assertion.Context,
+							Trace:   true,
+						})
+
+						if assertion.ErrorCode == 0 {
+							require.NoError(t, err, detailedInfo)
+							require.Equal(t, assertion.Expectation, resp.GetAllowed(), detailedInfo)
+						} else {
+							require.Error(t, err, detailedInfo)
+							e, ok := status.FromError(err)
+							require.True(t, ok, detailedInfo)
+							require.Equal(t, assertion.ErrorCode, int(e.Code()), detailedInfo)
+						}
+					})
+				}
+			})
+		}
+	})
+}
+
+func runTestMatrixSuite(t *testing.T, client ClientInterface) {
+	runTestMatrix(t, testParams{typesystem.SchemaVersion1_1, client})
```

Comment:
Understand that the focus is currently on check but where do you see LU and LO fitting into this matrix?

Reply:
Right now the types within this package are copy pasted across the different LO/LU/Check tests. And its just unmarshalling the yaml components it needs. In the future this suite would be moved to its own package to hold the model and states.

Then we can decide if LO/LU/Checks live side by side, or they are per package and each point to the respective stage. But that is the N+1.